### ISSUE TITLE
What is the rule for retries after a failed Storage.put, and how can I change it?

### ISSUE THREAD

**Issue by baharev** [aws-amplify/amplify-js#177]
TL;DR

- What is the rule for retries after a failed Storage.put?
- How can I change this rule? (number of retries, backoff, etc.)
- Why do I get a GET after a PUT that hasn't even been sent?
- Why do I get  `X-Amz-User-Agent: aws-amplify/0.1.22 js` even though I am using `0.1.32`?

-----

It seems to me that if a `Storage.put` fails, it will be immediately retried 3 times, meaning altogether 4 attempts before giving up. Here is a log of a failed `Storage.put`:

![screenshot_2018-01-21_00-15-15](https://user-images.githubusercontent.com/2605584/35193663-3c78b5f2-fea6-11e7-8518-c058fcf7b014.png)

It is nice that it retries automatically, but in my opinion it should be documented, and it should be configurable, such as the number of retries, backoff policy, etc. 

There are also a couple of strange things.

If the HTTP PUT fails in such a way that it hasn't even been sent (note the `(failed)` in red in the status column, and the provisional header), then I do not understand the GET request (`?max-keys=0`) that follows it. It seems pointless to me: The PUT hasn't been sent to the server. See the screenshots at the bottom of this post. Those images walk you through the request-response headers, and one image shows in detail the PUT that hasn't been sent. 

By the way, the PUT most likely fails due to an issue with Chrome. I suspect some same origin policy in the background: I selected an image in my Google Drive. This means that the PUT would have to upload an image from google.com to amazonaws.com, while the script's origin is cloudfront.net. I think Chrome gets confused, and gives the rather weird `ERR_FILE_NOT_FOUND` error. The same succeeds in Firefox for Android which tells me that the code is fine, and the file does exist. It is Chrome that gives the wrong error message.

The other strange thing is the `X-Amz-User-Agent: aws-amplify/0.1.22 js`. I am using 0.1.32. I grepped the source code is `node_modules/aws-amplify/dist/aws-amplify.js` and it does have:

```
exports.Constants = {
    userAgent: 'aws-amplify/0.1.22 js'
};
```

I deleted the `node_modules`, re-installed everything, including aws-amplify 0.1.32 (double-checked this in my package.json), but I still get  `X-Amz-User-Agent: aws-amplify/0.1.22 js` which looks strange to me.

------

The request-response headers for the failed Storage.put:

![screenshot_2018-01-21_00-12-47](https://user-images.githubusercontent.com/2605584/35193943-1c1a960e-feab-11e7-8491-6e47a30f18b4.png)

![screenshot_2018-01-21_00-13-28](https://user-images.githubusercontent.com/2605584/35193945-22110b74-feab-11e7-9aca-51b9e3922822.png)

![screenshot_2018-01-21_00-13-49](https://user-images.githubusercontent.com/2605584/35193946-268eace2-feab-11e7-9692-744ba67ca3ff.png)

![screenshot_2018-01-21_00-13-58](https://user-images.githubusercontent.com/2605584/35193952-2d060f20-feab-11e7-8f3c-9bad5f5e9e22.png)

![chrome_network_event](https://user-images.githubusercontent.com/2605584/35193953-30ddcfb6-feab-11e7-992d-b74f7aa0e5f7.png)

**Comment by mlabieniec** [comment]
@baharev we are working on more enhanced s3 / storage support.

What is the rule for retries after a failed Storage.put?
 > we currently don't have this built in outside of what the lower level aws-sdk does for you see 
[here](https://github.com/aws/aws-sdk-js/issues/7030) and [here](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property). we will be adding ability to customize these options further with Storage.configure. Would you mind giving some feedback around what you might like to see here as well?

How can I change this rule? (number of retries, backoff, etc.)
 > we don't currently have an interface to customize this yet. Can you provide some feedback around what you'd like to see here? See links above around what you can configure with the S3.upload method we are using under the covers.

Why do I get a GET after a PUT that hasn't even been sent?
> The GET is likely to check the status of the upload in the event of a failure, and to determine where to begin the retry using intelligent concurrent handling of parts if the payload is large enough.

Why do I get X-Amz-User-Agent: aws-amplify/0.1.22 js even though I am using 0.1.32?
> This was changed in around 0.1.34 i believe. Right now it defaults to 0.1.x and we will update this based on major releases.

**Comment by baharev** [comment]
@mlabieniec  I will answer in greater detail but until then:

> we currently don't have this built in outside of what the lower level aws-sdk does for you see
> here and here. we will be adding ability to customize these options further with Storage.configure. Would you mind giving some feedback around what you might like to see here as well?

The first link is broken, I get a 404, please check it.

> The GET is likely to check the status of the upload in the event of a failure, and to determine where to begin the retry using intelligent concurrent handling of parts if the payload is large enough.

It was a single upload, a `putObject`, and not a multipart upload.

> This was changed in around 0.1.34 i believe. Right now it defaults to 0.1.x and we will update this based on major releases.

OK. A bit confusing but OK. I thought my installation was messed up.

As for the rest, I will answer later.

**Comment by mlabieniec** [comment]
@baharev understood, the Storage.put in Amplify performs an upload method:
https://github.com/aws/aws-amplify/blob/master/packages/aws-amplify/src/Storage/Storage.ts#L161

is what I was referring to with the GET.

**Comment by baharev** [comment]
> we currently don't have this built in outside of what the lower level aws-sdk does for you see
> here and here. we will be adding ability to customize these options further with Storage.configure. Would you mind giving some feedback around what you might like to see here as well?

If the underlying `aws-sdk` is doing the retries, then the only thing that I would like to see is this feature documented (altogether 4 attempts will be made in case of a failure). I am not happy with it, but if the underlying SDK is doing it, then the `aws-amplify` library can't do much about it.

> we don't currently have an interface to customize this yet. Can you provide some feedback around what you'd like to see here? See links above around what you can configure with the S3.upload method we are using under the covers.

Ideally, I would want to manage the retries myself. The second best solution would be configuration (number of retries; delay and factor for the backoff).

> The GET is likely to check the status of the upload in the event of a failure, and to determine where to begin the retry using intelligent concurrent handling of parts if the payload is large enough.

I tried to locate this part of the code in the `aws-sdk` with grep but I have given up. In any case, the object was well below 5MB, meaning a single `putObject`.

**Comment by baharev** [comment]
OK, I think I have found it:

- [maxRetries](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#maxRetries-property)
- [retryDelayOptions](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#retryDelayOptions-property)

They apparently did the same that I suggested in the comment just above (number of retries; delay and factor for the backoff).

**Comment by baharev** [comment]
This weird behavior:

> If the HTTP PUT fails in such a way that it hasn't even been sent (note the (failed) in red in the status column, and the provisional header), then I do not understand the GET request (?max-keys=0) that follows it. It seems pointless to me: The PUT hasn't been sent to the server. See the screenshots at the bottom of this post. Those images walk you through the request-response headers, and one image shows in detail the PUT that hasn't been sent.

Is explained [here](https://github.com/aws/aws-sdk-js/issues/1895#issuecomment-360303394):

> The max-keys=0 request you're seeing is the S3 client attempting to determine the region in which the bucket being interacted with is located. The S3 client will often only see a generic "NetworkingError" in the browser, which prevents it from differentiating between legitimate networking errors (such as a laptop disconnecting from the internet) and the errors we would see if the request were sent to or signed for the wrong region.

**Comment by jordanranz** [comment]
I am closing this issue now as it seems the question has been answered. Please open a new issue if you have any further questions, feature requests, or bug reports.
